go/felix/intdataplane/endpoint_mgr_test.go

```diff
@@ -0,0 +1,406 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package intdataplane
+
+import (
+	log "github.com/Sirupsen/logrus"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/projectcalico/felix/go/felix/ipsets"
+	"github.com/projectcalico/felix/go/felix/iptables"
+	"github.com/projectcalico/felix/go/felix/proto"
+	"github.com/projectcalico/felix/go/felix/rules"
+	"github.com/projectcalico/felix/go/felix/set"
+	"reflect"
+)
+
+type mockTable struct {
+	expectedUpdate chan []*iptables.Chain
+	expectedRemove chan []*iptables.Chain
+
+	lastWorkDispatch []*iptables.Chain
+	lastHostDispatch []*iptables.Chain
+	lastHostEndpoint []*iptables.Chain
+}
+
+func newMockTable() *mockTable {
+	return &mockTable{
+		expectedUpdate: make(chan []*iptables.Chain, 2),
```

Comment:
If you're worried about policing for issues like reprogramming a chain unexpectedly, you can check for that in the UpdateChain method or you can record the changes in addition to the final state.